File: Python/DesignPattern/CreationalDesignPattern/FactoryMethodDesignPattern.py
# Factory Method Design Pattern
from abc import ABC, abstractmethod


# Product interface
class INotification(ABC):
    @abstractmethod
    def send(self): pass


class SMSNotification(INotification):
    def send(self): print("Sending SMS notification")


class WhatsAppNotification(INotification):
    def send(self): print("Sending WhatsApp notification")


class SlackNotification(INotification):
    def send(self): print("Sending Slack notification")


# Factory interface
class INotificationFactory(ABC):
    @abstractmethod
    def get_notification(self) -> INotification: pass


class SMSNotificationFactory(INotificationFactory):
    def get_notification(self) -> INotification:
        return SMSNotification()


class WhatsAppNotificationFactory(INotificationFactory):
    def get_notification(self) -> INotification:
        return WhatsAppNotification()


class SlackNotificationFactory(INotificationFactory):
    def get_notification(self) -> INotification:
        return SlackNotification()


class NotificationService:
    def send_notification(self, factory: INotificationFactory):
        notification = factory.get_notification()
        notification.send()


if __name__ == "__main__":
    service = NotificationService()
    service.send_notification(SMSNotificationFactory())
    service.send_notification(WhatsAppNotificationFactory())
    service.send_notification(SlackNotificationFactory())

File: C++/DesignPattern/CreationalDesignPattern/FactoryMethodDesignPattern.cpp
// Factory Method Design Pattern
#include <iostream>
#include <memory>
using namespace std;

// Product interface
class INotification {
public:
    virtual void send() = 0;
    virtual ~INotification() {}
};

class SMSNotification : public INotification {
public:
    void send() override { cout << "Sending SMS notification\n"; }
};

class WhatsAppNotification : public INotification {
public:
    void send() override { cout << "Sending WhatsApp notification\n"; }
};

class SlackNotification : public INotification {
public:
    void send() override { cout << "Sending Slack notification\n"; }
};

// Factory interface
class INotificationFactory {
public:
    virtual shared_ptr<INotification> getNotification() = 0;
    virtual ~INotificationFactory() {}
};

class SMSNotificationFactory : public INotificationFactory {
public:
    shared_ptr<INotification> getNotification() override {
        return make_shared<SMSNotification>();
    }
};

class WhatsAppNotificationFactory : public INotificationFactory {
public:
    shared_ptr<INotification> getNotification() override {
        return make_shared<WhatsAppNotification>();
    }
};

class SlackNotificationFactory : public INotificationFactory {
public:
    shared_ptr<INotification> getNotification() override {
        return make_shared<SlackNotification>();
    }
};

class NotificationService {
public:
    void sendNotification(INotificationFactory* factory) {
        auto notification = factory->getNotification();
        notification->send();
    }
};

int main() {
    NotificationService service;

    SMSNotificationFactory smsFactory;
    service.sendNotification(&smsFactory);

    WhatsAppNotificationFactory waFactory;
    service.sendNotification(&waFactory);

    SlackNotificationFactory slackFactory;
    service.sendNotification(&slackFactory);

    return 0;
}
